### ISSUE TITLE
REST API - no server-push updates if subscribed with transport websocket

### ISSUE THREAD

**Issue by chrwick** [openhab/openhab#2591]
Platform: Raspberry Pi 2 (Raspian) 
OpenHAB Version 1.6.2
Java(TM) SE Runtime Environment (build 1.8.0-b132)
jquery.atmosphere 2.2.11
tested with Firefox 37.0.2 as client
```
sitemap default label="Wohnung"
{
        Frame label="Klima" {
                Text item=temp_knx
                Text item=temp_out
                Text item=humid_out
                Text item=temp_in
                Text item=humid_in
        }
        Frame label="Misc" {
                Group          item=gAll
                Switch         item=hue_toggle_1                                label="Hue"
                Colorpicker    item=hue_color_1    visibility=[hue_toggle_1==ON]    label="Hue Farbe"
                Slider         item=hue_dimm_1     visibility=[hue_toggle_1==ON]    label="Hue Helligkeit"
                Text           item=door_entry
                Text           item=door_entry_sabotage
                Text           item=hm_switch_unreach
                Text           item=hm_switch_lowbat
        }
}
```

```
Group gAll

/* Hue items */
Switch hue_toggle_1   "left bulb"   <Switching> (gAll)    {hue="1"}
Color  hue_color_1    "left bulb"   <Colorize>      {hue="1"}
Dimmer hue_dimm_1     "left bulb"   <WhiteDimmer>   {hue="1;brightness;30"}

/* Homematic items */
Number temp_out "Temperatur Balkon oben [%.1f °C]"      <temperature> { homematic="address=EEQ0047170, channel=1, parameter=TEMPERATURE" }
Number humid_out "Luftfeuchtigkeit Balkon oben [%d %%]" <humidity> { homematic="address=EEQ0047170, channel=1, parameter=HUMIDITY" }
Number temp_in "Temperatur Wohnzimmer [%.1f °C]"      <temperature> { homematic="address=IEQ0145704, channel=1, parameter=TEMPERATURE" }
Number humid_in "Luftfeuchtigkeit Wohnzimmer [%d %%]" <humidity> { homematic="address=IEQ0145704, channel=1, parameter=HUMIDITY" }
Number door_entry "Wohnungstür [MAP(window.map):%d]" <contact>   {homematic="address=EEQ0049963, channel=1, parameter=STATE"}
String door_entry_sabotage "Wohnungstür [%s]" <contact>   {homematic="address=EEQ0049963, channel=1, parameter=ERROR"}

Switch hm_switch_unreach "Taster Signal Status [%s]" {homematic="address=EEQ0047942, channel=0, parameter=UNREACH"}
Switch hm_switch_lowbat "Taster Batterie Status [%s]" {homematic="address=EEQ0047942, channel=0, parameter=LOWBAT"}
Switch hm_switch_1 "Taster 1" {homematic="address=EEQ0047942, channel=1, parameter=PRESS_SHORT"}
Switch hm_switch_2 "Taster 2" {homematic="address=EEQ0047942, channel=2, parameter=PRESS_SHORT"}
Switch hm_switch_3 "Taster 3" {homematic="address=EEQ0047942, channel=3, parameter=PRESS_SHORT"}
Switch hm_switch_4 "Taster 4"  {homematic="address=EEQ0047942, channel=4, parameter=PRESS_SHORT"}

/* KNX items */
Number temp_knx "Temperatur KNX [%.1f °C]" <temperature>  { knx="3/1/0" }
```

Problem:
Subscription to /rest/sitemaps/default/default or /rest/items/gAll does not send any updates from items on the homepage (named "default") or the group (named "gAll"). 
Only the subscription to the state of one item sends updates.

jquery.atmosphere is connected successfully using websockets.

Code used running on external web server:
```javascript
                io.socket = $.atmosphere;

                var request = {
                                url: 'http://' + io.address + ':' + io.port + '/rest/sitemaps/default/default',
                                contentType : "application/json",
                                headers : { Accept: "application/json" },
                                logLevel : "debug",
                                transport : "websocket",
                                enableProtocol: false,
                                fallbackTransport: "long-polling"};

                request.onOpen = function(response) {
                        console.log("Atmosphere connected using " + response.transport);
                };

                request.onError = function(response) {
                        console.log("Atmosphere problem: " + response);
                };

                request.onMessage = function (response) {
                        var message = response.responseBody;
                        var item, val;
                        //var data = JSON.parse(message);
                        // DEBUG:
                        console.log("[io.openhab] receiving data: ", message);
```

The wiki says.
"You can subscribe on the following resources:

    a single item (or item group)
    an item state
    a single page of a sitemap"

So I consider the first page, the homepage as valid page.

**Comment by chrwick** [comment]
I have also tried the "Publish / Subscribe Test for the openHAB REST API" HTML test page.

It shows the same behaviour if websocket is selected as transport. 
It works for long-polling, streaming and jsonp though.

**Comment by Testato** [comment]
+1

**Comment by NickWaterton** [comment]
I have not been able to get this working using streaming either. Streaming only works on a single item, or a sitemap page (not the default page). I can't get anything out of websockets either.

Anyone have a working example? ie headers to send, items/sitemaps that work. I'm trying to get notifications on all items, I don't want to have to set up a connection for each item, or each page.

I'm using python for this btw.

Thanks,

**Comment by NickWaterton** [comment]
OK I have figured out the websockets issue, the headers have to be attached as a querystring to the url. You can't send them as regular headers.

In python websockets, this looks like:

client = WSClient()
client.connect("ws://192.168.100.119:8009/rest/sitemaps/nicks/Lights?Accept=application/json")

(where WSClient is a websocket class based on websockets).

This still only works on single items and sitemap pages though.

**Comment by digitaldan** [comment]
That's correct, but 1.8 will expand this to all items, see
https://github.com/openhab/openhab/pull/2973 , not sure if this meets your
needs.

On Wed, Sep 23, 2015 at 1:13 PM NickWaterton <notifications@github.com>
wrote:

> OK I have figured out the websockets issue, the headers have to be
> attached as a querystring to the url. You can't send them as regular
> headers.
>
> In python websockets, this looks like:
>
> client = WSClient()
> client.connect("ws://
> 192.168.100.119:8009/rest/sitemaps/nicks/Lights?Accept=application/json")
>
> (where WSClient is a websocket class based on websockets).
>
> This still only works on single items and sitemap pages though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openhab/openhab/issues/2591#issuecomment-142701120>.
>
-- 
Sent from my mobile.

**Comment by 9037568** [comment]
Closed via merged PR #2973.
